### PR TITLE
[web][api][SIMS #1341]Offering bulk upload - DB restrictions and offering UI date-only fix

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Fixed the date-only values on Offering UI which were displaying the day before instead of the correct day.
- Small improvements for Offering upload UI.

**All constraints/indexes below will execute a data check to ensure that the current data does not have duplications.**
- Created a DB constraint to ensure that SABC code presents on _sims.education_programs_ will be unique for an institution.
- Created a DB constraint to ensure that institution_code presents on _sims.institution_locations_ will be unique for an institution.
- Created a unique index for "Approved" and "Creation pending" offering to ensure that they will be unique for a location/program in terms of name and start and end dates.
**As talked with @abschwenker, new tickets will be created to handle to Uadditonal UI validations due to the inclusion of the above constraints.**

- Changed how the records are inserted. Now any error at the DB level will interrupt the entire process. This decision was made because after any command participating in the same transaction causes an error, the transaction is marked as aborted and all subsequent commands will fail, so there was no value in executing the process using the method as it was. If a recorded caused an error, it will be displayed as shown below.

image

- Changed CSV parser configuration to use `skipEmptyLines: "greedy"` to avoid parsing empty records or records having only white spaces.

### CODE REVIEW

File: sources/packages/api/src-sql/EducationPrograms/Add-sabc-code-unique-constraint.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+-- Remove existing codes to avoid conflics while creating the constraint.
```
Comment (ann-aot) [inline sources/packages/api/src-sql/EducationPrograms/Add-sabc-code-unique-constraint.sql:1]: typo `conflics`

File: sources/packages/api/src-sql/EducationPrograms/Add-sabc-code-unique-constraint.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+SET
+  sabc_code = NULL;
+
+-- Create the unique contraint for institution_id and sabc_code.
+ALTER TABLE
+  sims.education_programs
+ADD
+  CONSTRAINT institution_id_sabc_code_unique UNIQUE (institution_id, sabc_code);
+
+COMMENT ON CONSTRAINT institution_id_sabc_code_unique ON sims.education_programs IS 'Ensures SABC code will be unique for a particular institution. The uniqueness is critical to allow the offering bulk upload which uses the SABC code as a unique identifier for a program under a specific institution.';
```
Comment (ann-aot) [inline sources/packages/api/src-sql/EducationPrograms/Add-sabc-code-unique-constraint.sql:13]: 👍

File: sources/packages/api/src-sql/EducationProgramsOfferings/Add-offering-duplication-unique-index.sql [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        SELECT
+          location_id,
+          program_id,
+          offering_name,
+          study_start_date,
+          study_end_date
+        FROM
+          sims.education_programs_offerings
+        WHERE
+          offering_status IN ('Approved', 'Creation pending')
```
Comment (ann-aot) [inline sources/packages/api/src-sql/EducationProgramsOfferings/Add-offering-duplication-unique-index.sql:23]: why only ('Approved', 'Creation pending')

File: sources/packages/web/src/views/institution/OfferingsUpload.vue [hunk 4]
```diff
@@ -85,7 +85,7 @@
             class="ml-2"
             color="primary"
             prepend-icon="fa:fa-solid fa-upload"
-            :loading="loading"
+            :loading="creationProcessing"
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/institution/OfferingsUpload.vue:88]: The linear one is reporting the progress of the file upload only. It is the Axios progress, the same we use for form.io progress.
The circular one is the total loading of the operation, file upload + server processing, so the linear will be done and the button one will be still spinning for a while until API sends the response back.

File: sources/packages/api/src-sql/EducationPrograms/Add-sabc-code-unique-constraint.sql (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src-sql/EducationPrograms/Add-sabc-code-unique-constraint.sql]: Fixed.

File: sources/packages/web/src/views/institution/OfferingsUpload.vue [hunk 5]
```diff
@@ -107,8 +107,12 @@
...
         />
       </v-form>
-      <content-group v-if="showValidationSummary">
-        <p class="category-header-medium primary-color">Validation summary</p>
+      <content-group v-if="validationResults.length">
+        <body-header
+          title="Validation summary"
+          :recordsCount="validationResults.length"
+        ></body-header>
+        <p class="category-header-medium primary-color"></p>
```
Comment (ann-aot) [inline sources/packages/web/src/views/institution/OfferingsUpload.vue:115]: its an empty p tag

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/api/src/services/education-program-offering/education-program-offering-validation.models.ts [hunk 6]
```diff
@@ -610,13 +611,12 @@ export interface OfferingValidationResult {
  */
 export interface ValidatedOfferingInsertResult {
   validatedOffering: OfferingValidationResult;
-  insertResult: InsertResult;
+  insertResult?: InsertResult;
```
[the remark this answers is not in the document]
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/education-program-offering/education-program-offering-validation.models.ts]: This was a previous change that is no longer needed. I just removed the `ValidatedOfferingInsertResult` entirely.
Thanks for noticing it 😉

**Review by dheepak-aot — APPROVED** [review]
Nice work @andrewsignori-aot 👍

File: sources/packages/web/src/views/institution/OfferingsUpload.vue (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/views/institution/OfferingsUpload.vue]: Removed.

File: sources/packages/api/src/database/error-handler.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * Database constraints and indexes that must be used for further verifications.
+ */
+export enum DatabaseConstraintNames {
+  /**
+   * Ensures offerings in "Approved" or "Creation pending" statuses
+   * does not have the same name, study start, and study end dates for
+   * a specific location and program.
+   * Applied on table sims.education_programs_offerings.
+   */
+  LocationIDProgramIDOfferingNameStudyDatesIndex = "location_id_program_id_offering_name_study_dates_index",
```
Comment (guru-aot) [inline sources/packages/api/src/database/error-handler.ts:13]: The names are getting bigger and bigger :)

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Pull request merged**